### PR TITLE
[cuda-api-wrappers] new port

### PR DESCRIPTION
**Pull request by eclipse0922**
Upstream: https://github.com/eyalroz/cuda-api-wrappers

Related PR:  https://github.com/microsoft/vcpkg/pull/30888 
Thanks to @eladm-ultrawis

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

**Comment by eclipse0922** [comment]
`CMAKE_CUDA_ARCHITECTURES` is None on CI, but its valid on local  :crying_cat_face:

File: ports/cuda-api-wrappers/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+set(CUDA_ARCHITECTURES "native")
+
+vcpkg_cmake_configure(
+	SOURCE_PATH "${SOURCE_PATH}"
+	OPTIONS
+	-DCAW_BUILD_EXAMPLES=OFF
+	-DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES}
+    -DCMAKE_CUDA_COMPILER=${CMAKE_CUDA_COMPILER}
+	)
```
Comment (Adela0814) [inline ports/cuda-api-wrappers/portfile.cmake:30]:
```suggestion
vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
    OPTIONS
        -DCAW_BUILD_EXAMPLES=OFF
        -DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES}
        -DCMAKE_CUDA_COMPILER=${CMAKE_CUDA_COMPILER}
)
```

File: ports/cuda-api-wrappers/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,55 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO eyalroz/cuda-api-wrappers
+    REF v0.6.6
```
Comment (Adela0814) [inline ports/cuda-api-wrappers/portfile.cmake:4]:
```suggestion
    REF "v${VERSION}"
```

File: ports/cuda-api-wrappers/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	-DCAW_BUILD_EXAMPLES=OFF
+	-DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES}
+    -DCMAKE_CUDA_COMPILER=${CMAKE_CUDA_COMPILER}
+	)
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup(
+PACKAGE_NAME cuda-api-wrappers
+CONFIG_PATH lib/cmake
+)
```
Comment (Adela0814) [inline ports/cuda-api-wrappers/portfile.cmake:36]:
```suggestion
vcpkg_cmake_config_fixup(
    PACKAGE_NAME cuda-api-wrappers
    CONFIG_PATH lib/cmake
)
```

**Review by Adela0814 — CHANGES_REQUESTED** [review]

**Comment by Adela0814** [comment]
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

**Comment by eyalroz** [comment]
I'm not experienced with vcpkg, but I will note that I see an attempt to "find CUDA", but don't see a check for whether NVIDIA driver is installed. These two things are generally separate, i.e. you can have the CUDA toolkit without a driver and vice-versa. So perhaps that check is necessary as well. Or not, I can't say for sure.

**Comment by eclipse0922** [comment]
@Adela0814 Than you for the reivew. I fxed algiment.

@eyalroz 
```
find_program(NVIDIA_SMI "nvidia-smi")

if(NOT NVIDIA_SMI)
    message(STATUS "NVIDIA driver not found")
else()
    execute_process(COMMAND ${NVIDIA_SMI} --query-gpu=driver_version --format=csv,noheader
                    OUTPUT_VARIABLE NVIDIA_DRIVER_VERSION
                    OUTPUT_STRIP_TRAILING_WHITESPACE)
    message(STATUS "NVIDIA Driver Version: ${NVIDIA_DRIVER_VERSION}")
endif()
```
This snipper woudl work. I gues

**Comment by eclipse0922** [comment]
@eyalroz Driver checking make CI failure, 
I understand your concern. But I don't think you should put in a nvidia driver check.
Why not just assume that the developer PC has both?

File: ports/cuda-api-wrappers/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+# windows nvcc compiler path
+if (WIN32)
+    set(CMAKE_CUDA_COMPILER "${CUDA_TOOLKIT_ROOT}/bin/nvcc.exe")
+endif()
+
+# unix nvcc compiler path
+if (UNIX)
+    set(CMAKE_CUDA_COMPILER "${CUDA_TOOLKIT_ROOT}/bin/nvcc")
+endif()
```
Comment (dg0yt) [inline ports/cuda-api-wrappers/portfile.cmake:21]:
```suggestion
set(CMAKE_CUDA_COMPILER "${CUDA_TOOLKIT_ROOT}/bin/nvcc${VCPKG_HOST_EXECUTABLE_SUFFIX}")
```
or just put the expression after `-DCMAKE_CUDA_COMPILER=` below.

File: ports/cuda-api-wrappers/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+set(CUDA_ARCHITECTURES "native")
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DCAW_BUILD_EXAMPLES=OFF
+        -DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES}
+        -DCMAKE_CUDA_COMPILER=${CMAKE_CUDA_COMPILER}
```
Comment (dg0yt) [inline ports/cuda-api-wrappers/portfile.cmake:30]:
```suggestion
        "-DCMAKE_CUDA_ARCHITECTURES=${CUDA_ARCHITECTURES}"
        "-DCMAKE_CUDA_COMPILER=${CMAKE_CUDA_COMPILER}"
```

File: ports/cuda-api-wrappers/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    CONFIG_PATH lib/cmake
+)
+
+set(CAW_CMAKE_PACKAGE_FILES_DIR "${CURRENT_PACKAGES_DIR}/share/cuda-api-wrappers")
+
+file(GLOB packageFiles ${CAW_CMAKE_PACKAGE_FILES_DIR}/cuda-api-wrappers/*)
+foreach(pkgFile ${packageFiles})
+    get_filename_component(fileName ${pkgFile} NAME)
+    file(RENAME ${pkgFile} ${CAW_CMAKE_PACKAGE_FILES_DIR}/${fileName})
+endforeach()
```
Comment (dg0yt) [inline ports/cuda-api-wrappers/portfile.cmake:45]:
```suggestion
vcpkg_cmake_config_fixup(CONFIG_PATH share/cuda-api-wrappers)
```
or whatever the original config path really is.

File: ports/cuda-api-wrappers/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+file(REMOVE_RECURSE "${CAW_CMAKE_PACKAGE_FILES_DIR}/cuda-api-wrappers")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/lib" "${CURRENT_PACKAGES_DIR}/lib")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
```
Comment (dg0yt) [inline ports/cuda-api-wrappers/portfile.cmake:50]: Header-only? Then don't build debug. 
Add an early
~~~cmake
set(VCPKG_BUILD_TYPE release) # header-only
~~~

File: ports/cuda-api-wrappers/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)
```
Comment (dg0yt) [inline ports/cuda-api-wrappers/portfile.cmake:52]:
```suggestion
file(COPY "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
```

**Comment by eclipse0922** [comment]
@dg0yt  Thanks a lot!

**Comment by eyalroz** [comment]
> @eyalroz Driver check causes CI failure.

Really? A machine doesn't need to have an actual GPU in order to have the driver.

> I understand your concern. But I don't think we should put in a nvidia driver check. Why not just assume that the developer PC has both? Of course, as you say, the driver should work without the toolkit, and vice versa. To break it down like that, I think it would be too much to write all the features in the vcpkg port depending on the mode.

Well, the library is header-only - and thus doesn't really require anything else; but to you want to build something with the wrapper library e.g. its example programs - you have to have both the toolkit and the driver. 

In the library's CMakeLists, I require both the CUDA runtime library and the driver. Although the toolkit-provided stub is sufficient to satisfy that ... Anyway, I guess it's fine the way you have it now.

**Comment by eclipse0922** [comment]
@eyalroz These are the CI log when I tried to put driver check in CMakeLists.txt


Linux
https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/97109/artifacts?artifactName=failure%20logs%20for%20x64-linux&api-version=7.1&%24format=zip

Windows
https://dev.azure.com/vcpkg/c1ee48cb-0df2-4ab3-8384-b1df5a79fe53/_apis/build/builds/97109/artifacts?artifactName=failure%20logs%20for%20x64-windows&api-version=7.1&%24format=zip

```
find_program(NVIDIA_SMI "nvidia-smi")

if(NOT NVIDIA_SMI)
    message(FATAL_ERROR "NVIDIA driver not found")
else()
```
Above driver check works fine on local machine (both Windows and Linux)
But did not work on the CI machine
```
-- Using source at /mnt/vcpkg-ci/buildtrees/cuda-api-wrappers/src/v0.6.6-e9c3c9be44.clean
CMake Error at ports/cuda-api-wrappers/portfile.cmake:13 (message):
  NVIDIA driver not found
Call Stack (most recent call first):
  scripts/ports.cmake:170 (include)
```

**Comment by eclipse0922** [comment]
@eyalroz I mie
I think it is just path problem, I will figure it out.

**Comment by Adela0814** [comment]
Usage tested pass on `x64-windows`.

**Review by Adela0814 — DISMISSED** [review]

**Review by BillyONeal — APPROVED** [review]
I'm a bit nervous to give `cuda-Xxx` to something that isn't officially associated with CUDA but it's already on repology so this is OK.

**Comment by BillyONeal** [comment]
Thanks :)

**Comment by eclipse0922** [comment]
Thank you for the reviews.

File: versions/baseline.json [hunk 9]
```diff
@@ -2024,6 +2024,10 @@
       "baseline": "10.1",
       "port-version": 13
     },
+    "cuda-api-wrappers" : {
```
Comment (dg0yt) [inline versions/baseline.json:2027]: Please use `vcpkg x-add-version` for version record updates. It avoids non-standard formatting. (Corrections appearing in new PRs. Fortunately not causing merge conflicts, but may confuse reviewers.)

Comment (eclipse0922) [reply versions/baseline.json:2027]: Thanks for the clarification.

**Pull request merged**